src/graphics/vulkan.zig

```diff
@@ -156,3 +179,111 @@ pub fn createInstance() void {
 }
+
+// MARK: Physical Device
+
+const deviceExtensions = [_][*:0]const u8{
+	c.VK_KHR_SWAPCHAIN_EXTENSION_NAME,
+};
+
+const QueueFamilyIndidices = struct {
+	graphicsFamily: ?u32 = null,
+	presentFamily: ?u32 = null,
+
+	fn isComplete(self: QueueFamilyIndidices) bool {
+		return self.graphicsFamily != null and self.presentFamily != null;
+	}
+};
+
+fn findQueueFamilies(dev: c.VkPhysicalDevice) QueueFamilyIndidices {
+	var result: QueueFamilyIndidices = .{};
+	const queueFamilies = getPhysicalDeviceQueueFamilyProperties(main.stackAllocator, dev);
+	defer main.stackAllocator.free(queueFamilies);
+	for(queueFamilies, 0..) |family, i| {
+		if(family.queueFlags & c.VK_QUEUE_GRAPHICS_BIT != 0) {
+			result.graphicsFamily = @intCast(i);
+		}
+		var presentSupport: u32 = 0;
+		checkResult(c.vkGetPhysicalDeviceSurfaceSupportKHR(dev, @intCast(i), surface, &presentSupport));
+		if(presentSupport != 0) {
+			result.presentFamily = @intCast(i);
+		}
+	}
+	return result;
+}
+
+fn checkDeviceExtensionSupport(dev: c.VkPhysicalDevice) bool {
+	const availableExtension = enumerateDeviceExtensionProperties(main.stackAllocator, dev, null);
+	defer main.stackAllocator.free(availableExtension);
+	for(deviceExtensions) |requiredName| continueOuter: {
+		for(availableExtension) |available| {
+			if(std.mem.eql(u8, std.mem.span(requiredName), std.mem.span(@as([*:0]const u8, @ptrCast(&available.extensionName))))) {
+				break :continueOuter;
+			}
+		}
+		std.log.warn("Rejecting device because extension {s} was not found", .{requiredName});
+		return false;
+	}
+	return true;
+}
+
+fn deviceScore(dev: c.VkPhysicalDevice) f32 {
+	var properties: c.VkPhysicalDeviceProperties = undefined;
+	c.vkGetPhysicalDeviceProperties(dev, &properties);
+	var features: c.VkPhysicalDeviceFeatures = undefined;
+	c.vkGetPhysicalDeviceFeatures(dev, &features);
+	std.log.debug("Device: {s}", .{@as([*:0]const u8, @ptrCast(&properties.deviceName))});
+	std.log.debug("Properties: {}", .{properties});
+	std.log.debug("Features: {}", .{features});
+
+	const baseScore: f32 = switch(properties.deviceType) {
+		c.VK_PHYSICAL_DEVICE_TYPE_CPU => 1e-9,
+		c.VK_PHYSICAL_DEVICE_TYPE_DISCRETE_GPU => 1e9,
+		c.VK_PHYSICAL_DEVICE_TYPE_INTEGRATED_GPU => 1,
+		else => 0.1,
+	};
+
+	const availableExtension = enumerateDeviceExtensionProperties(main.stackAllocator, dev, null);
+	defer main.stackAllocator.free(availableExtension);
+	std.log.debug("Device extensions:", .{});
+	for(availableExtension) |ext| {
+		std.log.debug("\t{s}", .{ext.extensionName});
+	}
+	if(!findQueueFamilies(dev).isComplete() or !checkDeviceExtensionSupport(dev)) return 0;
+
+	if(features.multiDrawIndirect != c.VK_TRUE) {
+		std.log.warn("Rejecting device: multDrawIndirect is not supported", .{});
+		return 0;
+	}
+
+	if(features.dualSrcBlend != c.VK_TRUE) {
+		std.log.warn("Rejecting device: dual source blending is not supported", .{});
+		return 0;
+	}
+
+	return baseScore;
+}
+
+fn pickPhysicalDevice() !void {
```

Comment:
Do we want to give user some control of which GPU is chosen?
It would be useful in multi GPU systems and for testing purposes.